### PR TITLE
DPE-110 remove primary edge case

### PR DESCRIPTION
**Pull request by MiaAltieri**
### Problem

In short this PR address this JIRA ticket [DPE-110](https://warthogs.atlassian.net/jira/software/c/projects/DPE/boards/390?modal=detail&selectedIssue=DPE-110&quickFilter=844)

In long: 
- In a previous [PR](https://github.com/canonical/mongodb-operator/pull/23) it was mentioned that one of the integration tests was failing due to an edge case in removing primary units. This PR address this.
- The problem that was occurring was that when a primary unit got deleted MongoDB went into Split Brain, and could not recover.
- What was causing this problem to occur was that after the unit with the primary got removed from Juju, the remaining units were not able to reconfigure the replica set (even though a new primary had already been elected)
 
### Solution

As mentioned earlier the core problem was that: once the primary unit got deleted the remaining units could not reconfigure the replica set. So our solution was simple:
- Do not allow juju to remove the unit until it has reconfigured the replica set. 

But how was this done? It required a lot of structural changes in `src/charm.py` them being:
- in `src/charm.py` - `_on_mongodb_storage_detaching` gets triggered for the unit that gets deleted, here we do two things
     1. we step down the primary (if necessary)
     2. we reconfigure the replica set. By reconfiguring here, we address our core problem.
- in `src/charm.py` - `_on_mongodb_relation_departed` we **no longer** reconfigure here, we have now moved the responsibility of reconfiguring to the unit that is getting deleted. **BUT** we do update the relation data for `replica_set_hosts`, the reason being is that the leader unit must do this and so it might not be possible to do in  `_on_mongodb_storage_detaching` (if a non-leader is getting deleted).
-  in `src/charm.py` - `_on_leader_elected` we **no longer** reconfigure here, we have now moved the responsibility of reconfiguring to the unit that is getting deleted. **BUT** we do update the relation data for `replica_set_hosts`, the reason being is that if we have a new leader, only the leader can update these and after our units have departed we will have a better idea of what should go here

### Context

- "Split Brain": A split-brain scenario is the result of two data sets (which were originally synced) losing the ability to sync
- MongoDB must reconfigure the dataset before losing touch with the replica that is getting deleted.

### Release Notes

- `src/charm.py` - `_on_mongodb_storage_detaching` now responsible for stepping down primary and reconfiugring
- `src/charm.py` - `_on_mongodb_relation_departed` now only updates replica set hosts 
-  `src/charm.py` - `_on_leader_elected` now only updates replica set hosts 
-  `src/mongod_helpers.py` - `remove_replica` removes a specified IP from the configuration

### CODE REVIEW

File: src/mongod_helpers.py [hunk 1]
```diff
@@ -167,6 +165,40 @@ def initialise_replica_set(self, hosts: list) -> None:
...
+            logger.debug("current config for replica set: %s", current_members)
+            new_members = []
+
+            # add members to new config that aren't getting removed
+            for member in current_members:
+                # get member ip without ":PORT"
+                member_ip = member["host"].split(":")[0]
+
+                # if this ip is still being used retain it in new config
+                if not member_ip == remove_replica_ip:
```
Comment (WRFitch) [inline src/mongod_helpers.py:192]: This order of operations feels a little clunky - could you use != instead?

**Review by WRFitch — CHANGES_REQUESTED** [review]
Looks good, a couple of minor complaints

File: src/mongod_helpers.py [hunk 2]
```diff
@@ -131,8 +131,6 @@ def primary_step_down(self) -> None:
             client.admin.command("replSetStepDown", config)
         except (ConnectionFailure, ConfigurationError, OperationFailure) as e:
             raise e
-        finally:
-            client.close()
```
Comment (marceloneppel) [inline src/mongod_helpers.py:135]: Is it correct to remove this connection close call? (the other calls to `client` are calling `close`)

**Review by marceloneppel — COMMENTED** [review]
LGTM Mia! I left two comments about a call to `close` in the client connection. It's also important to check what Will commented about the duplicated observer call on `_on_leader_elected`.

File: src/mongod_helpers.py (hunk 1 above)
Comment (MiaAltieri) [reply src/mongod_helpers.py:192]: Ah, sorry for some reason I though I couldn't do `!=`

File: src/charm.py [hunk 3]
```diff
@@ -56,69 +55,73 @@ def __init__(self, *args):
...
+            event.defer()
             return
 
-        # do not allow leader to reconfigure the set if leader is getting removed
-        if departing_unit == self.unit.name:
-            # make note that that the new leader must reconfigure
-            self._peers.data[self.app]["_new_leader_must_reconfigure"] = "True"
-            return
+        # if a new leader has been elected, reconfigure the replica set
+        self.framework.observe(self.on.leader_elected, self._on_leader_elected)
```
[the remark this answers is not in the document]
Comment (MiaAltieri) [reply src/charm.py:113]: Agreed 👍

File: src/mongod_helpers.py (hunk 2 above)
Comment (MiaAltieri) [reply src/mongod_helpers.py:135]: Not sure how this got removed, defintely necessary, great catch :)

File: tests/unit/test_mongod_helpers.py [hunk 4]
```diff
@@ -270,3 +270,25 @@ def test_primary_failure(self, mock_client, client):
...
+            ConnectionFailure("error message"),
+            ConfigurationError("error message"),
+            OperationFailure("error message"),
+        ]
+        raises = [ConnectionFailure, ConfigurationError, OperationFailure]
+        for exception, expected_raise in zip(exceptions, raises):
+            with self.assertRaises(expected_raise):
+                mock_client.admin.command.side_effect = exception
+                mongo.primary_step_down()
+                mock_client.close.assert_called()
```
[the remark this answers is not in the document]
Comment (MiaAltieri) [reply tests/unit/test_mongod_helpers.py:294]: I see! I misunderstood how `assertRaises` works, thanks for this catch! 👁️

File: src/charm.py [hunk 5]
```diff
@@ -56,69 +55,73 @@ def __init__(self, *args):
...
-        # acquire removed unit
-        # TODO update this to use ops framework after this issue is addressed:
-        # https://github.com/canonical/operator/issues/707
-        # this will be updated in the following PR
-        departing_unit = os.environ.get("JUJU_DEPARTING_UNIT")
-
-        # If primary is leaving, allow it to step down before reconfiguring
-        if departing_unit == self.unit.name and departing_unit == self._primary:
+        logger.debug("the current primary is %s", self._primary)
+        logger.debug("the current unit is %s", self.unit.name)
```
Comment (zmraul) [inline src/charm.py:79]: Maybe join them into just one log?

**Review by WRFitch — APPROVED** [review]
Looks great!

File: src/mongod_helpers.py [hunk 6]
```diff
@@ -167,6 +165,40 @@ def initialise_replica_set(self, hosts: list) -> None:
...
+                member_ip = member["host"].split(":")[0]
+
+                # if this ip is still being used retain it in new config
+                if not member_ip == remove_replica_ip:
+                    new_members.append(member)
+
+            rs_config["config"]["members"] = new_members
+            replica_set_client.admin.command("replSetReconfig", rs_config["config"])
+        except (ConfigurationError, ConfigurationError, OperationFailure) as e:
+            raise e
```
Comment (zmraul) [inline src/mongod_helpers.py:200]: Since `e` is not being used because you are handling the exceptions on other methods, you can leave it as:
```
except (ConfigurationError, ConfigurationError, OperationFailure):
    raise
```
Also, you could just omit it, but this is done so it's more clear what exceptions are being handled?

**Review by zmraul — APPROVED** [review]
LGTM! A couple of annoying nitpicks :)

File: src/mongod_helpers.py (hunk 6 above)
Comment (MiaAltieri) [reply src/mongod_helpers.py:200]: Exactly, I wanted to make it clear about which exceptions are being handled :) , but also I wanted to ensure that even if these exceptions occur that we execute the code in the `finally:` block

I didn't realise I could remove the `as e:`, TIL :) Thanks Raul :D

**Review by delgod — APPROVED** [review]

**Review by marceloneppel — APPROVED** [review]
